File: sdk/javascript/src/exceptions.ts
export class SafeRunError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "SafeRunError";
  }
}

export class SafeRunAPIError extends SafeRunError {
  statusCode: number;
  constructor(statusCode: number, message: string) {
    super(`SafeRun API error (${statusCode}): ${message}`);
    this.name = "SafeRunAPIError";
    this.statusCode = statusCode;
  }
}

export class SafeRunApprovalTimeout extends SafeRunError {
  changeId: string;
  timeoutMs: number;
  constructor(changeId: string, timeoutMs: number) {
    super(`Timed out waiting for approval of change ${changeId} after ${timeoutMs}ms`);
    this.name = "SafeRunApprovalTimeout";
    this.changeId = changeId;
    this.timeoutMs = timeoutMs;
  }
}

File: sdk/python/saferun/exceptions.py
"""Custom exceptions for SafeRun SDK."""

class SafeRunError(Exception):
    """Base exception for SafeRun SDK."""


class SafeRunAPIError(SafeRunError):
    """Raised when API returns an error response."""

    def __init__(self, status_code: int, message: str):
        super().__init__(f"SafeRun API error ({status_code}): {message}")
        self.status_code = status_code
        self.message = message


class SafeRunApprovalTimeout(SafeRunError):
    """Raised when waiting for approval times out."""

    def __init__(self, change_id: str, timeout: int):
        super().__init__(f"Timed out waiting for approval of change {change_id} after {timeout}s")
        self.change_id = change_id
        self.timeout = timeout
